File: Retos/Reto #35 - PRIMEROS PASOS [Fácil]/go/blackriper.go
package main

import (
	"fmt"
	"strconv"
)

/* Hola mundo*/ 

func main(){
  fmt.Println("Hello World!!")
}

/* Tipos de datos */ 
var message string ="Hola como estas?"
var microcode rune ='h' // o byte
var precio int=12 // o  int8 int16 int32 int64
var taxes  float32=12.5
var cost  float64= 135.56
var send  bool = false

/* declaracion de variables */
// constante
const fileName string="Archivo.pdf"
// variable global
var isFriday bool=false

/* estructuras condicionales*/ 
var ninja string ="Naruto"
// tiene que ir dentro de la funcion main
if ninja=="Naruto"{
  fmt.Printf("%v ha usando el Rasenchuriken!!",ninja)
}else{
  fmt.Println("No ninja !!")
}

/* Estructuras de datos */

// slices o array
var hokages []string=[]{"Hashirama","Tobirama","Hiruzen","Minato"}

// diccionarios o mapas 
var ninjaTools map[string]string

ninjaTools["shuriken"]="shuriken"
ninjaTools["papelboom"]="papel bomba"

/*Estructuras de repeticion*/

//for in
for index,hokages:= range hokages{
  fmt.Printf("%v fue el %v hokage de Konoha",hokage,index)
}

// for como while
 var countdown= 10 
 for countdown>0{
   fmt.Println(countdown--)
 }

// for tradicional 

for indx:=0;indx<len(hokages);indx++{
  fmt.Println(hokages[indx])
}

/* funciones */ 

func NinjaJutsu(){
  fmt.Println("Sasuke usa el jutsu bola de fuego")
}

func IsBijuu(name string) bool{
 if name=="Kyubii"{
   return true
 }else if name =="Shukaku"{
   return true
 }
 return false
}

/* go no utliza clases pero hace algo parecido */ 

// la  estructura struct sirve para definir tipos de datos complejos en este caso es como definir los atributos de una clase 
type Shinobi struct{
  Name string
  Clan string
  Jutsu string
}
// las funciones reciben una propiedad extra llamada Reciver  estas le indican que este metodo pertenece a la estruct 
func (s *Shinobi) NewJutsu(jutsu string){
  s.Jutsu=jutsu
}

func (s Shinobi) Attack(){
   fmt.Printf("%v usa el jutsu %v",s.Name,s.Jutsu)
 }

func ejemplo(){
 shikamaru:= &Shinobi{Name:"shikamaru",Clan:"Nara",Jutsu:"atadura de sombra"}
 shikamaru.Attack()
 shikamaru.NewJutsu("Clon de sombra")
 shikamaru.Attack()
 }

 /* Control de errores*/ 
  func ControlError(){
    var price string = "124"
    // convertir variable a entero
    num,err:=strconv.Atoi(price)
    
    // hacer algo en caso de error 
    if err!=nil{
      panic(err)
    }
    fmt.Println(num)
  }

File: Retos/Reto #35 - PRIMEROS PASOS [Fácil]/typescript/blackriper.ts

/* hola mundo*/

// arrow funcion
const hello = () => console.log("Hello World")

// funcion tradicional
function Hello() {
  console.log("Hello World")
}

/* tipos de datos */
let numero: number = 10
let message: string = "Hola"
let send: boolean = false || true
let big: bigint = 100n;
let u: undefined = undefined;
let n: null = null;
let strictlyTyped: unknown = 4;
let looselyTyped: any = 4;

/* declaracion de variables*/

// constante 
const filename = "texto.txt"
// variable global 
var precio = 12.5
// variable local 
let tecnica = "Kaioken"

/* estructuras condicionales */
let warrior = "Goku"
if (warrior == "Goku") {
  console.log(`${warrior} uso el kamehame-ha`)
} else {
  console.log(`${warrior} uso vuelo`)
}

/* estructuras de datos*/
//array o list
let warriors = ["Goku", "Vegeta", "Kriin", "Gohan", "Picoro", "Trunks"]
// tupla 
let tecnicas: [string, number] = ["Kaioken", 20]
//set 
let villanos = new Set<string>(["Frezzer", "Cell", "MajinBoo"])

// object 
interface Warrior {
  name: string;
  race: string;
  age: number;
}

const goku: Warrior = {
  name: "SonGoku",
  race: "Sayan",
  age: 39
}

/* estructuras de repeticion*/

warriors.forEach((warrior) => console.log(warrior))

for (let index = 0; index < villanos.size; index++) {
  console.log(villanos[index])
}

let countdown = 10

while (countdown > 0) {
  console.log(countdown--)
}

/*funciones*/

function AttackWarrior(): string {
  return "Picoro utilizo el Makankossapo!!"
}

function Attack(name: string, tecnica: string) {
  console.log(`${name} utilizo el ${tecnica}`)
}

/* clases */
class Guerrero {
  private nombre: string;
  private ataque: string;

  constructor(nom: string, atk: string) {
    this.nombre = nom;
    this.ataque = atk;
  }

  Attack() {
    console.log(`${this.nombre} utilizo el ${this.ataque}`)
  }
}
// instanciar clase para crear un object warrior y acceder a sus metodos
let gotenks = new Guerrero("Gotenks", "Fantasmas Kamikaze!!")
gotenks.Attack()

// control de errores 
const cuentame_un_chiste = async () => {
  try {
    const res = await fetch("https://api.chucknorris.io/jokes/random")
    const data = await res.json()
    console.log(data.value)
  } catch (err) {
    console.log(err)
  }
}













